src/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -145,8 +145,8 @@ public static unsafe void Copy(Array sourceArray, Array destinationArray, int le
             MethodTable* pMT = RuntimeHelpers.GetMethodTable(sourceArray);
             if (pMT == RuntimeHelpers.GetMethodTable(destinationArray) &&
                 !pMT->IsMultiDimensionalArray &&
-                (uint)length <= (uint)sourceArray.Length &&
-                (uint)length <= (uint)destinationArray.Length)
+                (uint)length <= (nuint)sourceArray.LongLength &&
+                (uint)length <= (nuint)destinationArray.LongLength)
```

Comment:
Does working with longs here provide a performance penalty on 32bit platforms such as x86?

Reply:
I assume the JIT will ignore the `int -> long -> uint` conversion as a no-op, but I don't know that for sure. But since we're talking SzArrays here, wouldn't we want to use `uint` in all cases anyway since we know we're talking 2 bn elements max?

Reply:
I have followed the pattern used in existing code.

I will check what this does to code quality. Maybe we should just have NativeLength property on the array and convert to native ints early in cases like these to avoid going back and forth.

Reply:
The `NativeLength` property would save 4 bytes of code in this method on x86 (missed JIT optimization) and 3 bytes of native code on x64. Not worth it.

Reply:
Looks good to me. I was worried the jit might do something terrible, but hoping for a result like you found.

Reply:
+1 for `NativeLength`. Will fit nicely when C# makes `nint` first class type. Perhaps can be an internal extension method on `Array` for starters.

Reply:
+1 for `NativeLength`. Will fit nicely when C# makes `nint` first class type. Perhaps can be an internal extension method on `Array` for starters.